PlateSwipe/PlateSwipe
# Enhance Scanner UI and Tolerance of OFF Json Parsers

# Enhance Scanner UI and Tolerance of OFF Json Parsers
## Description
This pull request includes changes to improve the handling of ingredient fetching and display in the camera scanning and fetch functionalities. The most important changes include the introduction of new state flows to track fetching states, refactoring of ingredient display components, and enhancements to the ingredient parsing logic.

### Improvements to ingredient fetching and display:

* `IngredientViewModel.kt`: Added `_isFetchingByBarcode` state flow to track barcode fetching state and renamed `_isSearching` to `_isFetchingByName` for clarity. This change was also done in the interface it implements.
* `CameraScanCodeBarScreen.kt`: Refactored `IngredientOverlay` to be more modularized and readable. It also now handle 2 extra case. It used to only be able to show the found ingredient, but now it has a screen for when no ingredient that has been found and when it's searching for said ingredient.
I also made sure the overlay would only pop up when we did the first search.

New Loading Screen:
![Screenshot 2024-11-28 160901](https://github.com/user-attachments/assets/ca0d00ad-f15d-4aec-abc3-6f20b2a925ea)

New No Ingredient Found Screen:
![image](https://github.com/user-attachments/assets/fd1a7add-6c26-4d21-a1ea-3003bdbd3c21)

### Enhancements to ingredient parsing:

* `OpenFoodFactsIngredientRepository.kt`: Introduced `parseProductName` method to handle multiple language suffixes when parsing ingredient names. We used to miss alot of ingredients, because the field "product_name" wouldn't be present or would be empty. Often there would be an alternative ex: "product_name_en" or "product_name_de", so now we look for those too.

### Test updates:

* `CameraScanCodeBarScreenTest.kt`: Added `waitForIdle` to ensure UI stability, especially for the `isFetchingByBarcode` to update the UI.

## Testing Changes

1. Go to the add ingredient screen of create recipe. 
2. Then select the barcode scanner.
3. See if an overly appears before you scan something.
4. Scan an ingredient and see if there is a loading screen before it appears.
5. Scan an ingredient that isn't in openfoodfacts and check if you get a "No ingredient found message".

You can scan this image for a non-openfoodfacts ingredient:
![image](https://github.com/user-attachments/assets/7ba81e6e-ee2a-40cc-a6e9-cf70a8ac61e7)

## Checklist
- [x] All CI checks passed.
- [x] Linked issue/feature (if applicable): #247
